### PR TITLE
Add axis and where arguments to loss functions.

### PR DESCRIPTION
**Pull request by carlosgmartin**
#902

### CODE REVIEW

**Comment by vroulet** [comment]
Thanks @carlosgmartin, could you add tests?
Also you will need to wait for #916 to pass.

**Comment by fabianp** [comment]
@carlosgmartin , there are some conflicts with main, do you mind updating the pull request? thanks

**Comment by carlosgmartin** [comment]
@fabianp Done.

@vroulet Also added tests.

**Comment by carlosgmartin** [comment]
@vroulet @fabianp Let me know if you'd like me to make any other changes.

**Comment by fabianp** [comment]
hey @carlosgmartin , thanks for the ping and apologies for the late reply, most of the team is on vacation 🏖️ 

The addition of the axis argument looks good to me, and the use of an axis kwarg if fairly common in numpy-like functions. ✅ 

Regarding the "where" argument however, I haven't  seen it yet in other libraries. Do you know of any numpy/jax/flax/etc. functions that admit a "where" or a "mask" or similar kwarg? I just want to make sure that our API is as similar as possible to other libraries that have already implemented similar functionality.

**Comment by carlosgmartin** [comment]
Most reduction functions in numpy/jax take a `where` argument for masking. For example:

- [max](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.max.html)
- [min](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.min.html)
- [sum](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.sum.html)
- [mean](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.mean.html)
- [all](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.all.html)
- [any](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.any.html)
- [prod](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.prod.html)
- [var](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.var.html)
- [std](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.std.html)
- [softmax](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softmax.html)

Conceptually, it makes sense that any reduction function should take both `axis` and `where` arguments.

**Comment by fabianp** [comment]
excellent, thanks for the info!

**Comment by fabianp** [comment]
one more thing: could you please add the tag `.. versionchanged:: 0.2.4` to the docstrings of the functions you've changed explaining the change? See for example here for an example: https://github.com/google-deepmind/optax/blob/main/optax/schedules/_inject.py#L114

**Comment by fabianp** [comment]
also, the new kwargs should be described in the function docstring (on top of adding the `versionchanged` tag)

**Comment by carlosgmartin** [comment]
@fabianp Done.

**Comment by fabianp** [comment]
thanks, we're almost there. Please add type annotations for these two new kwargs. These are likely 

```
where: chex.Array | None = None,
axis: int | tuple[int, ...] | None = -1
```

**Comment by carlosgmartin** [comment]
@fabianp Done.

**Review by fabianp — APPROVED** [review]
thanks!
